stable_baselines3/common/recurrent/buffers.py

```diff
@@ -115,62 +161,109 @@ def __init__(
         buffer_size: int,
         observation_space: spaces.Space,
         action_space: spaces.Space,
-        hidden_state_shape: Tuple[int, int, int, int],
+        hidden_state_example: TensorTree,
         device: Union[th.device, str] = "auto",
         gae_lambda: float = 1,
         gamma: float = 0.99,
         n_envs: int = 1,
     ):
-        self.hidden_state_shape = hidden_state_shape
-        self.seq_start_indices, self.seq_end_indices = None, None
-        super().__init__(buffer_size, observation_space, action_space, device, gae_lambda, gamma, n_envs)
+        super(RolloutBuffer, self).__init__(buffer_size, observation_space, action_space, device=device, n_envs=n_envs)
+        self.hidden_state_example = hidden_state_example
+        self.gae_lambda = gae_lambda
+        self.gamma = gamma
+
+        batch_shape = (self.buffer_size, self.n_envs)
+        device = self.device
+
+        self.observation_space_example = space_to_example((), observation_space)
+
+        self.advantages = th.zeros(batch_shape, dtype=th.float32, device=device)
+        self.returns = th.zeros(batch_shape, dtype=th.float32, device=device)
+        self.data = RecurrentRolloutBufferData(
+            observations=space_to_example(batch_shape, self.observation_space, device=device, ensure_non_batch_dim=True),
+            actions=th.zeros(
+                (*batch_shape, self.action_dim),
+                dtype=th.long if isinstance(self.action_space, (spaces.Discrete, spaces.MultiDiscrete)) else th.float32,
+                device=device,
+            ),
+            rewards=th.zeros(batch_shape, dtype=th.float32, device=device),
+            episode_starts=th.zeros(batch_shape, dtype=th.bool, device=device),
+            values=th.zeros(batch_shape, dtype=th.float32, device=device),
+            log_probs=th.zeros(batch_shape, dtype=th.float32, device=device),
+            hidden_states=tree_map(
+                lambda x: th.zeros((self.buffer_size, x.shape[0], self.n_envs, *x.shape[1:]), dtype=x.dtype, device=device),
+                hidden_state_example,
+            ),
+        )
+
+    # Expose attributes of the RecurrentRolloutBufferData in the top-level to conform to the RolloutBuffer interface
+    @property
+    def episode_starts(self) -> th.Tensor:  # type: ignore[override]
+        return self.data.episode_starts
+
+    @property
+    def values(self) -> th.Tensor:  # type: ignore[override]
+        return self.data.values
+
+    @property
+    def rewards(self) -> th.Tensor:  # type: ignore[override]
+        assert self.data.rewards is not None, "RecurrentRolloutBufferData should store rewards"
```

Comment:
Why do we assert here, but not equivalently in `episode_starts` and `values`?

Reply:
I completely forgot. Maybe rewards could be None before? It's not in the type annotation anymore so I removed the asserts.